### PR TITLE
`Exam mode`: Fix an issue with problem statement updates when an exercise was not opened

### PR DESCRIPTION
**Pull request by coolchock**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Motivation and Context

Currently, the problem statement of an exam exercise is only updated if the exercise had previously been opened. This might lead to confusion, because students will see outdated problem statement if hadn't opened the exercise before.

### Description

Fixes https://github.com/ls1intum/Artemis/issues/9009

Thanks to these changes, the problem statement of an exercise will be updated regardless of whether a student has opened the exercise or not. If a student had previously opened the exercise, the problem statement will be updated, and the differences between the old and the new problem statements will be highlighted. If they hadn't opened the exercise yet, the problem statement will still be updated, but the differences won't be highlighted. This is because the ExamExerciseUpdateHighlighterComponent is only initialized when the exercise component is initialized. However, we don't want to initialize all exercise components at once, as this could result in up to 16,000 REST calls to the backend simultaneously in the case of large exams, potentially overloading the system.

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam

1. Log in to Artemis as student
2. Log in to Artemis as instructor in different browser/incognito tab
3. Start exam as student, but do not open any exercise
4. Switch to instructor account, go to exam management and make some changes to the problem statement in one of the exercises of the exam
5. Switch to student account and check if an event was received and changes to the problem statement were reflected.
6. Switch to instructor account again, and make some changes to the problem statement in the same exercise.
7. Switch to student account and check if an event was received,changes to the problem statement were reflected and differences are highlighted.

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

File: src/test/javascript/spec/component/exam/participate/exam-participation.component.spec.ts [hunk 1]
```diff
@@ -885,44 +884,6 @@ describe('ExamParticipationComponent', () => {
     });
 
     describe('handleHandInEarly', () => {
-        it('should reset pageComponentVisited after the hand-in-early window is closed', () => {
```
Comment (sarpsahinalp) [inline src/test/javascript/spec/component/exam/participate/exam-participation.component.spec.ts:888]: If I understand correctly this case is not tested anymore?

**Review by sarpsahinalp — CHANGES_REQUESTED** [review]
One of the e2e tests is failing probably due to changes:
```
FAILURE: ExamResults.spec.ts:184:13 Check exam modelling exercise results

  [chromium] › e2e/exam/ExamResults.spec.ts:184:13 › Exam Results › Check exam exercise results › Check exam modelling exercise results 

    Retry #1 ───────────────────────────────────────────────────────────────────────────────────────

    Error: Timed out finding an element matching the "locator('#exercise-166').locator('#exercise-result-score')" locator

(23 more lines...)
```

**Review by sarpsahinalp — CHANGES_REQUESTED** [review]
Test report coverage seems to be failing as well

File: src/test/javascript/spec/component/exam/participate/exam-participation.component.spec.ts (hunk 1 above)
Comment (coolchock) [reply src/test/javascript/spec/component/exam/participate/exam-participation.component.spec.ts:888]: yes, it is not needed anymore

**Review by muradium — APPROVED** [review]
Tested on TS4, received announcement before opening any exercise

**Review by EneaGore — APPROVED** [review]
Test on TS4. Recieved Notifications before and after opening exercises

**Review by BBesrour — DISMISSED** [review]
Tested in a testing session. Also code lgtm

**Review by rstief — APPROVED** [review]
Tested in testing session, got notifications when on the exam page before doing anything (as intended)

**Review by pzdr7 — DISMISSED** [review]
Code

**Comment by krusche** [comment]
There was a specific reason why exercises were only initialized when they are visited for the first time. I don't think we can simply remove that. I think it is related to the fallback when the instructor forgets to click on prepare exercise start

**Review by krusche — CHANGES_REQUESTED** [review]
Let's discuss if removing `pageComponentVisited` can lead to issues

**Review by JohannesStoehr — COMMENTED** [review]
Does this lead to increased rest calls at the start of the exam if all components are loaded directly?
If so, this could be a problem with 2000 students participating at the same time

**Comment by coolchock** [comment]
> Does this lead to increased rest calls at the start of the exam if all components are loaded directly? If so, this could be a problem with 2000 students participating at the same time

No, all student exam related information (exercises, participations, submissions, etc.) are fetched when user starts an exam. They are then passed as inputs to the components that are initialised.

**Review by julian-christl — COMMENTED** [review]
As far as I analysed, this is what happened before:
When a user first opened the page, no exercise was loaded. Only when they opened one exercise the first time, onInit was executed once.
When a user clicked on hand in early and then cancelled, all exercises but the one they were to return to were marked as "not visited" causing the page they return to not to execute onInit, however, once visiting other pages, they would retrigger onInit there.

Now, we trigger OnInit on all pages when starting first and when cancelling hand in early, just nothing happens in regards to onInit.

I think that makes sense, the only concern would be the content of the OnInit and if it would put heavy load on the system. Assuming an exercise number of 10, that wouldn't be an issue. In case there are day-long-exams with many more exercises, this could be an issue on opening, but that's it.

**Comment by krusche** [comment]
I still think the change is problematic. When onInit is called on a programming exercise, a REST call will be executed as far as I know. In an exam with 8 programming exercises and 2.000 students, this would mean 16.000 REST calls during the synchronized exam start vs. 2.000 REST calls in the previous implementation. Of course, you won't notice this in a testing session with 1 exercise in an exam and 2-3 people opening the exam at different times, however this is not realistic!!!

**Review by pzdr7 — APPROVED** [review]
Tested on TS3. 

Created an exam + did not prepare the exercise start.

Changed the problem statement of a programming exercise before opening -> problem statement was correct when I opened the page. ✅ 

Changed it after opening -> was able to toggle between diff and new version ✅ 

(same for Text exercise)

----

I tried the same for an exam where I prepared the exercise start -- works as expected! ✅

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me

**Review by Strohgelaender — APPROVED** [review]
code

**Review by florian-glombik — APPROVED** [review]
tested on ts3

* Created an exam + and prepared the exercise start
* Changed the problem statement of the following exercises
     * programming
     * text
     * modeling
     * file-upload
* Update was reflected in the problem statement

I think that the solution is an improvement to the previous version, however, I do not think that the current solution is optimal, imagine the following scenario:
1. complete exercise 1 
2. reload the page and open exercise 2
3. now update the problem statement of exercise 1
4. the changes will not be highlighted and might be hard to sport

(If a reasonable description was supplied it might be easier to sport, however, when I tested the notification feature I completely missed that you can write a custom message - I am not sure if we want to rely on the instructor supplying a descriptive message there)

Considering that this is an edge case (which updating the problem statement during the exam should be anyways) and another solution leads to overhead the current solution seems fine for now.
However, we might want to refactor the Highlighting logic after the semester to support the mentioned scenario(highlighting changes for not yet loaded exercises aswell) ?

**Review by krusche — APPROVED** [review]

**Pull request merged**